address_translators.go

```diff
@@ -0,0 +1,53 @@
+package gocql
+
+import (
+	"log"
+	"net"
+)
+
+// AddressTranslator provides a way translate node addresses (and ports) when one is
+// discovered or received as a node event. This is especially useful in ec2 (when
+// using the EC2MultiRegionAddressTranslator) to translate public IPs to private IPs
+// when possible.
+type AddressTranslator interface {
+	// Translate will translate the provided address and/or port when a host is discovered.
+	Translate(addr string, port int) (string, int)
+}
+
+type AddressTranslatorFunc func(addr string, port int) (string, int)
+
+func (fn AddressTranslatorFunc) Translate(addr string, port int) (string, int) {
+	return fn(addr, port)
+}
+
+// IdentityTranslator will do nothing but return what it was provided. It is essentially a no-op.
+func IdentityTranslator() AddressTranslator {
+	return AddressTranslatorFunc(func(addr string, port int) (string, int) {
+		return addr, port
+	})
+}
+
+// EC2MultiRegionAddressTranslator will do a reverse DNS lookup on the address provided,
+// then do a DNS lookup on the first hostname found in order to potentially get the
+// private IP of the given address. If no results are found in either lookup,
+// the given address is returned. Port is always returned unchanged. Build with the
+// gocql_debug tag to see the address translation at work.
+func EC2MultiRegionAddressTranslator() AddressTranslator {
```

Comment:
Hey @Zariel, I'm not sure what you mean by doing this outside of Cassandra. It directly relates to configuring a ring for multi-region support in ec2. You'll find similar support in the [Java driver](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/policies/EC2MultiRegionAddressTranslator.java) and the [NodeJS driver](https://github.com/datastax/nodejs-driver/blob/master/lib/policies/address-resolution.js#L76), for instance. If you need it, I can describe in more detail our ring setup spread across 6 regions, but there are plenty of other people who have described this issue.